File: 2265-count-nodes-equal-to-average-of-subtree/2265-count-nodes-equal-to-average-of-subtree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def averageOfSubtree(self, root:TreeNode) -> int:
        result = [0]

        def dfs(node):
            if not node:
                return 0, 0
            left_sum, left_count = dfs(node.left)
            right_sum, right_count = dfs(node.right)
            total_sum = left_sum + right_sum + node.val
            total_count = 1 + right_count + left_count
            if total_sum // total_count == node.val:
                result[0]+=1
            return total_sum, total_count

        dfs(root)
        return result[0]

File: 2265-count-nodes-equal-to-average-of-subtree/2265-count-nodes-equal-to-average-of-subtree.c
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     struct TreeNode *left;
 *     struct TreeNode *right;
 * };
 */
int countNodes(struct TreeNode* root) {
    if (!root) return 0;
    return 1 + countNodes(root->left) + countNodes(root->right);
}

int totalSum(struct TreeNode* root) {
    if (!root) return 0;
    return root->val + totalSum(root->left) + totalSum(root->right);
}

int averageOfSubtree(struct TreeNode* root) {
    if (!root) return 0;
    int sum = root->val + totalSum(root->left) + totalSum(root->right);
    int totalNodes = 1 + countNodes(root->left) + countNodes(root->right);

    int avg = sum/totalNodes;
    if (avg == root->val) {
        return 1 + averageOfSubtree(root->left) + averageOfSubtree(root->right);
    }
    return averageOfSubtree(root->left) + averageOfSubtree(root->right);
}